### PR TITLE
Reimplement vk.com resolver from scratch. Now requires authorisation and...

### PR DESCRIPTION
**Pull request by theli-ua**
... is much more reliable

### CODE REVIEW

**Comment by theli-ua** [comment]
Potentially this vk.com resolver will have issues unless captcha support is available

**Comment by theli-ua** [comment]
I mean, it works without captcha support, but whenever request rate exceeds some threshold - it'll stop resolving requiring captcha input. This self recovers with time.

File: vkontakte/content/metadata.json [hunk 1]
```diff
@@ -1,11 +1,11 @@
 {
-    "name": "VKontakte",
+    "name": "VK.com",
     "pluginName": "vkontakte",
-    "author": "Krzysztof Klinikowski",
-    "email": "kkszysiu@gmail.com",
-    "version": "0.1",
+    "author": "Anton Romanov",
+    "email": "theli.ua@gmail.com",
+    "version": "0.0.1",
```
Comment (muesli) [inline vkontakte/content/metadata.json]: This is a bit problematic since we're decreasing the version here. Should rather bump to 0.2, I guess?

**Comment by willstott101** [comment]
Could you implement some sort of queue? Same as the login system in the TIDAL resolver. Except you'd need to automatically reject any promises which exceed the time-out.

**Comment by theli-ua** [comment]
I messed something up with that defered timeout call .... I still get an error
"undefined is not an object (evaluating 'params['access_token'] = this._access_token')" "" 
I understand I messed up scopes but still not sure how to properly fix this

**Comment by dschmidt** [comment]
I'm working on a few things (not high priority sadly) that will allow us to have a test suite

File: vkontakte/content/contents/code/vkontakte.js [hunk 2]
```diff
@@ -1,398 +1,291 @@
...
- *   Tomahawk is distributed in the hope that it will be useful,
- *   but WITHOUT ANY WARRANTY; without even the implied warranty of
- *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
- *   GNU General Public License for more details.
- *
- *   You should have received a copy of the GNU General Public License
- *   along with Tomahawk. If not, see <http://www.gnu.org/licenses/>.
+ * You should have received a copy of the CC0 Public Domain Dedication
+ * along with this software. If not, see:
+ * http://creativecommons.org/publicdomain/zero/1.0/
```
Comment (dschmidt) [inline vkontakte/content/contents/code/vkontakte.js]: I'd prefer a BSD/MIT license over creative commons for code, easier to handle license combinations with those

File: vkontakte/content/contents/code/vkontakte.js [hunk 3]
```diff
@@ -1,398 +1,291 @@
...
+        if (changed) {
+            this.init();
+        }
+    },
 
-                for(var i=0;i<audio_data.length; i++){
-                    //if (debugMode == true) { console.log(audio_data[i], Math.abs(parseInt(audio_data[i].duration) - duration), artist, song, audio_data[i].artist.toLowerCase(), artist.toLowerCase() == audio_data[i].artist.toLowerCase()); }
+    testConfig: function (config) {
+        return this._getLoginPromise(config).catch(function (error) {
+            throw new Error('Invalid credentials');
```
Comment (dschmidt) [inline vkontakte/content/contents/code/vkontakte.js]: Why don't you simply throw this from the `_getLoginPromise` method itself? :)

File: vkontakte/content/contents/code/vkontakte.js [hunk 4]
```diff
@@ -1,398 +1,291 @@
...
 
-                callback(audio_data)
-            } else {
-                callback({error:'not_found'})
-            }
-        })
+        if (!this._email || !this._password) {
+            Tomahawk.reportCapabilities(TomahawkResolverCapability.NullCapability);
+            //This is being called even for disabled ones
+            //throw new Error( "Invalid configuration." );
```
Comment (dschmidt) [inline vkontakte/content/contents/code/vkontakte.js]: Ack, this is a flaw in the current js plugin system in Tomahawk, unfortunately nothing we can change quickly.

File: vkontakte/content/contents/code/vkontakte.js [hunk 5]
```diff
@@ -1,398 +1,291 @@
...
+                    return results;
+                })
     },
-    resolve: function( qid, artist, album, title )
-    {
-        var data = {
-            qid: qid,
-            results: [],
+
+    _defer: function (callback, args, scope) {
```
Comment (dschmidt) [inline vkontakte/content/contents/code/vkontakte.js]: What is this defer logic good for?

File: vkontakte/content/contents/code/vkontakte.js (hunk 2 above)
Comment (theli-ua) [inline vkontakte/content/contents/code/vkontakte.js]: CC0 is a "public domain" as in "The person who associated a work with this deed has dedicated the work to the public domain by waiving all of his or her rights to the work worldwide under copyright law, including all related and neighboring rights, to the extent allowed by law. "

Isn't that more permissive than BSD/MIT ?

Comment (dschmidt) [inline vkontakte/content/contents/code/vkontakte.js]: I don't know, I'm not a lawyer and that's exactly why I'm asking for a different license. For BSD/MIT and (A/L)GPL licenses there are comparison and compatibility tables out there, I doubt you usually find creative commons licenses in there.

Comment (willstott101) [inline vkontakte/content/contents/code/vkontakte.js]: https://www.gnu.org/philosophy/license-list.html#CC0

File: vkontakte/content/contents/code/vkontakte.js [hunk 6]
```diff
@@ -1,398 +1,339 @@
...
-    } else {
-        if(url.match(/\?/))
-            xhr.open(method, url+'&'+data, true)
-        else
-            xhr.open(method, url+'?'+data, true)
+    logged_in: null, // null, = not yet tried, 0 = pending, 1 = success, 2 = failed
 
-        xhr.send()
-    }
+    _queue : {}, //we'll queue up resolve requests and execute them in batches
```
Comment (dschmidt) [inline vkontakte/content/contents/code/vkontakte.js]: You are later on looping over this with for in, then you need to either remember to check hasOwnProperty on every iteration, or you need to initialize this with `Object.create(null)`

File: vkontakte/content/contents/code/vkontakte.js [hunk 7]
```diff
@@ -1,398 +1,339 @@
...
+    },
 
-            // User logged off from vkontakte
-            if(xhr.responseText.match(/progress\.gif/)){
-                delete VK.search_method
+    newConfigSaved: function() {
+        var config = this.getUserConfig();
 
-                VK.search(artist, song, duration, callback)
+        var changed =
```
Comment (dschmidt) [inline vkontakte/content/contents/code/vkontakte.js]: style

File: vkontakte/content/contents/code/vkontakte.js [hunk 8]
```diff
@@ -1,398 +1,339 @@
...
-                  title: title.textContent.trim(),
-                  duration: duration,
-                  url: url
-                })
+        params['access_token'] = this._access_token;
+        params['v']            = this.API_VERSION;
+
+        return Tomahawk.get("https://api.vk.com/method/" + api, {
+            data: params
+        }).then(
```
Comment (dschmidt) [inline vkontakte/content/contents/code/vkontakte.js]: style

File: vkontakte/content/contents/code/vkontakte.js [hunk 9]
```diff
@@ -1,398 +1,339 @@
...
+            var params = {
+                count: limit || 300,
+                q: query,
+            };
+
+            return this._apiCall('audio.search',params).then( function (response) {
+                return response.response.items.map(that._convertTrack, trackInfo);
+            });
+        } else { //Array of queries
+            var code = 'return [' + query.map(function(q, i) { return 'API.audio.search({"q":"'+q+'","count":'+limit[i]+'})'; }).join(',') + '];';
```
Comment (dschmidt) [inline vkontakte/content/contents/code/vkontakte.js]: wtf?!

File: vkontakte/content/contents/code/vkontakte.js [hunk 10]
```diff
@@ -1,398 +1,339 @@
...
+                that._queue[qid] = [];
+            }
+            Tomahawk.log(resolve);
+            that._queue[qid].push(resolve);
+            Tomahawk.log(JSON.stringify(that._queue));
+        });
 
-                VK.search_method = response.search_method
+        if (!(this._batching)) {
+            setTimeout(function(){ that._batchResolve(that); }, 1000);
```
Comment (dschmidt) [inline vkontakte/content/contents/code/vkontakte.js]: Have you checked what timeout tomahawk has for waiting on results from a resolver?

**Comment by dschmidt** [comment]
I have a very bad gut feeling about this resolver, it's only kept together by lots and lots of hacks and duct tape.

I could merge it after you changed the license and fixed the style issues, but to make this clear: I won't ever maintain this piece of code and I'm voting strongly against shipping it by default via synchrotron  :P

File: vkontakte/content/contents/code/vkontakte.js (hunk 10 above)
Comment (muesli) [inline vkontakte/content/contents/code/vkontakte.js]: @dschmidt It will wait for however long the resolver told the pipeline to.

Comment (dschmidt) [inline vkontakte/content/contents/code/vkontakte.js]: Ah uh

File: vkontakte/content/contents/code/vkontakte.js (hunk 8 above)
Comment (theli-ua) [inline vkontakte/content/contents/code/vkontakte.js]: Could you elaborate on those style comments please?
I haven't seen any code style guidelines for resolvers in the repo so it is not clear to me which ones I violated. Thanks.

File: vkontakte/content/contents/code/vkontakte.js (hunk 9 above)
Comment (theli-ua) [inline vkontakte/content/contents/code/vkontakte.js]: This is no longer used, I'll just clean it up

**Comment by theli-ua** [comment]
I fell like its way less hacky than, say gmusic and 4shared you have in synchrotron, this one has pretty small not very complicated code base and uses a well documented and supported API while those rely on some reverse-engineering hacks.

File: vkontakte/content/contents/code/vkontakte.js (hunk 8 above)
Comment (dschmidt) [inline vkontakte/content/contents/code/vkontakte.js]: yep, no new line after then(

**Comment by dschmidt** [comment]
You are right, it's not that bad. I'm a bit stressed out in the last days and annoyed by doing a lot of context switching, didnt mean to be hostile. I'm sorry.

Path forward: 
1) Fix that object creation that you are looping on afterwards
2) Remove the odd newlines
3) check if you really want to keep all those `Tomahawk.log`s (if it makes sense to keep them, it's ok)
4) I'll merge

P.S.: I would still prefer a proper code license. I still haven't understood your issue with BSD/MIT

**Pull request merged**